File: server/src/shared/push.ts
import type { Db } from '../db.ts'
import { sb } from '../sb.ts'

interface PushRow { endpoint: string; p256dh: string; auth: string }

export async function notifyFarmMembers(
  db: Db,
  params: {
    farmId: string
    orderId: string
    type: 'order_created' | 'deposit_confirmed' | 'shipment_requested'
    title: string
    body: string
    url?: string
  },
): Promise<void> {
  const client = sb(db)
  const { data: members } = await client.from('farm_members').select('user_id').eq('farm_id', params.farmId)
  const userIds = [...new Set((members ?? []).map((row: any) => row.user_id as string))]
  if (userIds.length === 0) return

  await client.from('notifications').insert(
    userIds.map((userId) => ({
      user_id: userId,
      farm_id: params.farmId,
      order_id: params.orderId,
      type: params.type,
      title: params.title,
      body: params.body,
    })),
  )

  const { data: subs } = await client
    .from('push_subscriptions').select('endpoint, p256dh, auth').in('user_id', userIds)
  await sendPush((subs ?? []) as PushRow[], {
    title: params.title,
    body: params.body,
    url: params.url ?? `/admin/farms/${params.farmId}/orders`,
  })
}

export async function sendPush(
  subscriptions: PushRow[],
  payload: { title: string; body: string; url?: string },
): Promise<void> {
  const vapidPublic = process.env.VAPID_PUBLIC_KEY
  const vapidPrivate = process.env.VAPID_PRIVATE_KEY
  if (!vapidPublic || !vapidPrivate || subscriptions.length === 0) return

  try {
    const webpush = (await import('web-push')).default
    webpush.setVapidDetails('mailto:hello@farmassi.kr', vapidPublic, vapidPrivate)
    await Promise.allSettled(
      subscriptions.map((sub) =>
        webpush.sendNotification(
          { endpoint: sub.endpoint, keys: { p256dh: sub.p256dh, auth: sub.auth } },
          JSON.stringify(payload),
        ),
      ),
    )
  } catch (error) {
    console.error('web-push 실패', error)
  }
}

File: server-py/app/shared/push.py
import json
import os
from typing import Any, Iterable

import asyncpg

from ..sb import sb


async def notify_farm_members(
    conn: asyncpg.Connection,
    *,
    farm_id: str,
    order_id: str,
    type_: str,
    title: str,
    body: str,
    url: str | None = None,
) -> None:
    client = sb(conn)
    members = (await client.from_("farm_members").select("user_id").eq("farm_id", farm_id)).data
    user_ids = list(dict.fromkeys(row["user_id"] for row in (members or [])))
    if not user_ids:
        return

    await client.from_("notifications").insert([
        {"user_id": uid, "farm_id": farm_id, "order_id": order_id,
         "type": type_, "title": title, "body": body}
        for uid in user_ids
    ])

    subs = (await client.from_("push_subscriptions")
            .select("endpoint, p256dh, auth").in_("user_id", user_ids)).data
    await send_push(subs or [], title=title, body=body,
                    url=url or f"/admin/farms/{farm_id}/orders")


async def send_push(
    subscriptions: Iterable[dict[str, Any]], *, title: str, body: str, url: str | None = None
) -> None:
    vapid_public = os.environ.get("VAPID_PUBLIC_KEY")
    vapid_private = os.environ.get("VAPID_PRIVATE_KEY")
    subscriptions = list(subscriptions)
    if not vapid_public or not vapid_private or not subscriptions:
        return

    try:
        import asyncio

        from pywebpush import webpush

        payload = json.dumps({"title": title, "body": body, "url": url})

        def one(sub: dict[str, Any]) -> None:
            webpush(
                subscription_info={
                    "endpoint": sub["endpoint"],
                    "keys": {"p256dh": sub["p256dh"], "auth": sub["auth"]},
                },
                data=payload,
                vapid_private_key=vapid_private,
                vapid_claims={"sub": "mailto:hello@farmassi.kr"},
            )

        # pywebpush 는 동기라 이벤트 루프를 막지 않게 스레드로 돌린다.
        await asyncio.gather(
            *(asyncio.to_thread(one, sub) for sub in subscriptions), return_exceptions=True
        )
    except Exception as error:  # noqa: BLE001
        print("web-push 실패", error)
